### PR TITLE
fix(republish): filter out nodes with existing volume targets

### PR DESCRIPTION
**Pull request by tiagolobocastro**
Ensures we don't attempt to create nexus on a node which already has a nexus for a given volume.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

### CODE REVIEW

File: control-plane/grpc/proto/v1/volume/volume.proto [hunk 1]
```diff
@@ -150,6 +150,8 @@ message GetVolumesRequest {
   }
   // pagination to allow for multiple requests to get all volumes
   common.Pagination pagination = 2;
+  // ignore 404 not found errors
```
Comment (Abhinandan-Purkait) [inline control-plane/grpc/proto/v1/volume/volume.proto:153]: Question: Why is this needed?

File: control-plane/agents/src/bin/core/controller/scheduling/mod.rs [hunk 2]
```diff
@@ -92,6 +92,13 @@ impl NodeFilters {
             true
         }
     }
+    /// Should only attempt to use node where there are no targets for the current volume.
+    pub(crate) fn no_targets(request: &GetSuitableNodesContext, item: &NodeItem) -> bool {
+        let volume_targets = request.registry().specs().get_volume_nexuses(&request.uuid);
+        !volume_targets
+            .into_iter()
+            .any(|n| &n.lock().node == item.node_wrapper().id())
```
Comment (avishnu) [inline control-plane/agents/src/bin/core/controller/scheduling/mod.rs:100]: 1. Why is locking needed here? If it's needed, should it be on the 'volume_targets' instead of per item?
2. Is there a possibility that there is only one healthy node at the moment and a previous nexus for the same volume on the same node is in a shutdown/deletion state? Should we not select the node in that case? OR will it happen in the future reconciliation?

Comment (tiagolobocastro) [reply control-plane/agents/src/bin/core/controller/scheduling/mod.rs:100]: 1. Because that's how things are atm (irrespective of this PR), each  target lives on its own and has its own lock
2. Yes it's possible, in such a case we need something to clean up shutdown targets, which will be done as part of @abhilashshetty04 's PR, which will detect when we run out of nodes and clean up unused shutdown targets.

File: control-plane/grpc/proto/v1/volume/volume.proto (hunk 1 above)
Comment (tiagolobocastro) [reply control-plane/grpc/proto/v1/volume/volume.proto:153]: While debugging a cluster I noticed we end up with a ton of error messages which are actually not an error because on initial creation we do Get Volume xyz to find out if it exists or not, and first time it won't exist.
This prevents the error..
Usually we wouldn't need this flag but it seems we use the same grpc call to get 1 or many, so that's why it's needed as we cannot differentiate..

File: control-plane/agents/src/bin/core/controller/scheduling/mod.rs (hunk 2 above)
Comment (avishnu) [reply control-plane/agents/src/bin/core/controller/scheduling/mod.rs:100]: > 1. Because that's how things are atm (irrespective of this PR), each  target lives on its own and has its own lock

Why do we need to n.lock() here? Just trying to understanding.

Comment (tiagolobocastro) [reply control-plane/agents/src/bin/core/controller/scheduling/mod.rs:100]: Because the object in case is protected by a Mutex, eg: `Mutex<Obj>`, so to access its inner value (Obj) you have to take the lock which gives you a guard over it,eg: `Guard<Obj>`. When the guard is dropped, the mutex is unlocked.

Comment (avishnu) [reply control-plane/agents/src/bin/core/controller/scheduling/mod.rs:100]: I see, thanks for the explanation !!

File: control-plane/grpc/proto/v1/volume/volume.proto (hunk 1 above)
Comment (Abhinandan-Purkait) [reply control-plane/grpc/proto/v1/volume/volume.proto:153]: RIght. Thank you.

**Review by Abhinandan-Purkait — APPROVED** [review]

**Comment by tiagolobocastro** [comment]
bors merge
